homeassistant/components/sensor/filter.py

```diff
@@ -277,6 +288,46 @@ def _filter_state(self, new_state):
         return filtered
 
 
+@FILTERS.register(FILTER_NAME_TIME_SMA)
+class TimeSMAFilter(Filter):
+    """Simple Moving Average (SMA) Filter.
+
+    The window_size is determined by time, and SMA is time weighted.
+    """
+
+    def __init__(self, window_size, precision, entity):
+        """Initialize Filter."""
+        super().__init__(FILTER_NAME_TIME_SMA, 0, precision, entity)
+        self._time_window = int(window_size.total_seconds())
+        self.last_leak = None
+        self.queue = deque()
+
+    def _leak(self, now):
+        """Remove timeouted elements."""
+        while self.queue:
+            timestamp, _ = self.queue[0]
+            if timestamp + self._time_window <= now:
+                self.last_leak = self.queue.popleft()
+            else:
+                return
+
+    def _filter_state(self, new_state):
+        now = int(datetime.now().timestamp())
```

Comment:
Use our date util package and UTC timestamps:

```python
from homeassistant.util import date as dt_util

dt_util.utcnow()
```

That way you won't get hit by daylight savings etc.

Any reason you're using timestamps instead of datetime objects?

window size is already given to you in a timedelta. So you can do things like `dt_util.utcnow() - self._time_window` to calculate points in time.

Reply:
Time needs to be multiplied with an unknown type **value**, so I preferred to keep safe and make sure time is an int.